services/kafka/config.go

```diff
@@ -82,58 +84,59 @@ type WriteTarget struct {
-		return kafka.WriterConfig{}, errors.New("topic must not be empty")
+		return cfg, errors.New("topic must not be empty")
 	}
-	var balancer kafka.Balancer
+	var partitioner kafka.PartitionerConstructor
 	if target.PartitionById {
 		switch target.PartitionAlgorithm {
-		case "crc32c", "":
-			balancer = &kafka.CRC32Balancer{}
+		case "crc32", "":
+			// can't use kafka.NewCustomHashPartitioner, because we need to maintain compatibility
+			partitioner = newCRCPartitioner
 		case "murmur2":
-			balancer = &kafka.Murmur2Balancer{}
+			// can't use kafka.NewCustomHashPartitioner here either, because we need to maintain compatibility
+			partitioner = newMurmur2
+		case "murmur3":
+			partitioner = kafka.NewCustomHashPartitioner(murmur3.New32)
 		case "fnv-1a":
-			balancer = &kafka.Hash{}
+			partitioner = kafka.NewHashPartitioner
 		default:
-			return kafka.WriterConfig{}, fmt.Errorf("invalid partition algorithm: %q", target.PartitionAlgorithm)
+			return cfg, fmt.Errorf("invalid partition algorithm: %q", target.PartitionAlgorithm)
 		}
-	} else {
-		balancer = &kafka.LeastBytes{}
+		cfg.Producer.Partitioner = partitioner
 	}
 
-	var tlsCfg *tls.Config
+	cfg.Producer.Return.Errors = true // so we can display errors
+	cfg.Producer.Return.Successes = false
+	cfg.ClientID = c.ID
+	cfg.Metadata.Full = false // we only want to grab metadata for the topics we care about
+
 	if c.UseSSL {
-		t, err := tlsconfig.Create(c.SSLCA, c.SSLCert, c.SSLKey, c.InsecureSkipVerify)
+		var err error
+		cfg.Net.TLS.Enable = true
+		cfg.Net.TLS.Config, err = tlsconfig.Create(c.SSLCA, c.SSLCert, c.SSLKey, c.InsecureSkipVerify)
 		if err != nil {
-			return kafka.WriterConfig{}, err
+			return cfg, err
 		}
-		tlsCfg = t
 	}
-	dialer := &kafka.Dialer{
-		Timeout: time.Duration(c.Timeout),
-		TLS:     tlsCfg,
+	if c.Timeout > 0 {
+		cfg.Net.DialTimeout = time.Duration(c.Timeout)
+		cfg.Net.WriteTimeout = time.Duration(c.Timeout)
+		cfg.Net.ReadTimeout = time.Duration(c.Timeout)
 	}
 
-	baseConfig := kafka.WriterConfig{
-		Brokers:      c.Brokers,
-		Dialer:       dialer,
-		ReadTimeout:  time.Duration(c.Timeout),
-		WriteTimeout: time.Duration(c.Timeout),
-		BatchSize:    c.BatchSize,
-		BatchTimeout: time.Duration(c.BatchTimeout),
-		// Async=true allows internal batching of the messages to take place.
-		// It also means that no errors will be captured from the WriteMessages method.
-		// As such we track the WriteStats for errors and report them with Kapacitor's normal diagnostics.
-		Async: true,
-		ErrorLogger: kafka.LoggerFunc(func(s string, x ...interface{}) {
-			diagnostic.Error("kafka client error", fmt.Errorf(s, x...))
-		}),
-		Topic:    target.Topic,
-		Balancer: balancer,
+	cfg.Producer.Flush.MaxMessages = c.BatchSize
+	if cfg.Producer.Flush.MaxMessages <= 0 {
+		cfg.Producer.Flush.MaxMessages = DefaultBatchSize
 	}
+	cfg.Producer.Flush.Frequency = time.Duration(c.BatchTimeout)
 
-	return baseConfig, nil
+	// SASL
+	c.SASLAuth.SetSASLConfig(cfg)
```

Comment:
This can return an error.  Probably better to return that than to hope that `Validate` returns the same error.